### PR TITLE
Port MultiResourceRepository from platform/tools/adt/idea

### PR DESCRIPTION
**Pull request by kevinzheng-ap**
Part of https://github.com/cashapp/paparazzi/issues/524

cc @jrodbx

### CODE REVIEW

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/Renderer.kt [hunk 1]
```diff
@@ -66,17 +69,36 @@ internal class Renderer(
             }.apply { loadResources() }
           )
       } else {
-        val resourceFolderRepositories = environment.libraryResourceDirs.map {
+        val localResources = environment.localResourceDirs.map {
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/Renderer.kt:72]: if we're going to migrate more than just ResourceFolderRepository and AarSourceRepository, then let's revert these changes to Renderer and SessionParamsBuilder in this PR, because we'll need to then port Project/App/etc., which is fine, cuz I have it already ported!

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/LocalResourceRepository.kt [hunk 2]
```diff
@@ -14,6 +15,19 @@ import com.google.common.collect.ListMultimap
...
+    if (parents == null) {
+      parents = ArrayList(2) // Don't expect many parents
+    }
+    parents!!.add(parent)
+  }
+
+  open fun removeParent(parent: MultiResourceRepository) {
+    parents?.remove(parent)
+  }
+
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/LocalResourceRepository.kt:30]:
```suggestion
```

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/LocalResourceRepository.kt [hunk 3]
```diff
@@ -32,4 +46,13 @@ abstract class LocalResourceRepository protected constructor(
...
   ): Collection<ResourceItem> = throw UnsupportedOperationException("Not implemented yet")
+
+  /**
+   * Package accessible version of [.getMap].
+   * Do not call outside of [MultiResourceRepository].
+   */
+  open fun getMapPackageAccessible(
+    namespace: ResourceNamespace,
+    type: ResourceType
+  ): ListMultimap<String, ResourceItem>? = getMap(namespace, type)
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/LocalResourceRepository.kt:57]:
```suggestion
```

not needed

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/LocalResourceRepository.kt (hunk 2 above)
Comment (jrodbx) [reply paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/LocalResourceRepository.kt:30]: not needed

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+ * Ported from: [MultiResourceRepository.java](https://cs.android.com/android-studio/platform/tools/adt/idea/+/55991b4380c1ac18e81151493f59228220c5b72a:android/src/com/android/tools/idea/res/MultiResourceRepository.java)
+ *
+ * A super class for several of the other repositories. Its only purpose is to be able to combine
+ * multiple resource repositories and expose it as a single one, applying the “override” semantics
+ * of resources: earlier children defining the same resource namespace/type/name combination will
+ * replace/hide any subsequent definitions of the same resource.
+ *
+ * <p>In the resource repository hierarchy, MultiResourceRepository is an internal node, never a leaf.
+ */
+class MultiResourceRepository(displayName: String) : LocalResourceRepository(displayName) {
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt:54]:
```suggestion
abstract class MultiResourceRepository internal constructor(displayName: String) : LocalResourceRepository(displayName) {
```

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+  private var localResources = listOf<LocalResourceRepository>()
+
+  private var libraryResources = listOf<AarSourceResourceRepository>()
+
+  /** A concatenation of [.localResources] and [.libraryResources].  */
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt:59]:
```suggestion
  /** A concatenation of [localResources] and [libraryResources].  */
```

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+
+  private var resourceComparator =
+    ResourceItemComparator(ResourcePriorityComparator(ImmutableList.of()))
+
+  private val cachedMaps = ResourceTable()
+
+  /** Names of resources from local leaf repositories.  */
+  private val resourceNames: Table<SingleNamespaceResourceRepository, ResourceType, Set<String>> =
+    Tables.newCustomTable(
+      HashMap<SingleNamespaceResourceRepository, Map<ResourceType, Set<String>>>()
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt:78]:
```suggestion
      HashMap()
```

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt [hunk 7]
```diff
@@ -0,0 +1,535 @@
...
+    ) { Maps.newEnumMap(ResourceType::class.java) }
+
+  /** Describes groups of resources that are out of date in [.cachedMaps].  */
+  private val unreconciledResources: Table<ResourceNamespace, ResourceType, Set<SingleNamespaceResourceRepository>> =
+    Tables.newCustomTable(
+      HashMap<ResourceNamespace, Map<ResourceType, Set<SingleNamespaceResourceRepository>>>()
+    ) { Maps.newEnumMap(ResourceType::class.java) }
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt:85]:
```suggestion
```

not needed

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt [hunk 8]
```diff
@@ -0,0 +1,535 @@
...
+
+  fun setChildren(
+    localResources: List<LocalResourceRepository>,
+    libraryResources: Collection<AarSourceResourceRepository>,
+    otherResources: Collection<ResourceRepository>
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt:90]:
```suggestion
```

not needed for now

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt [hunk 9]
```diff
@@ -0,0 +1,535 @@
...
+  ) {
+    for (child in localResources) {
+      child.removeParent(this)
+    }
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt:94]:
```suggestion
```

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt [hunk 10]
```diff
@@ -0,0 +1,535 @@
...
+    computeLeafs(this, mapBuilder)
+    leafsByNamespace = mapBuilder.build()
+    mapBuilder = ImmutableListMultimap.builder()
+    computeNamespaceMap(this, mapBuilder)
+    repositoriesByNamespace = mapBuilder.build()
+    resourceComparator =
+      ResourceItemComparator(ResourcePriorityComparator(leafsByNamespace.values()))
+    for (child in localResources) {
+      child.addParent(this)
+    }
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt:111]:
```suggestion
this.children = buildList(this.localResources.size + this.libraryResources.size) {
      addAll(this@MultiResourceRepository.localResources)
      addAll(this@MultiResourceRepository.libraryResources)
    }

    leafsByNamespace =
      ImmutableListMultimap.builder<ResourceNamespace, SingleNamespaceResourceRepository>()
        .apply { computeLeafs(this@MultiResourceRepository, this) }
        .build()

    repositoriesByNamespace =
      ImmutableListMultimap.builder<ResourceNamespace, SingleNamespaceResourceRepository>()
        .apply { computeNamespaceMap(this@MultiResourceRepository, this) }
        .build()

    resourceComparator =
      ResourceItemComparator(ResourcePriorityComparator(leafsByNamespace.values()))
```

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt [hunk 11]
```diff
@@ -0,0 +1,535 @@
...
+      return getResources(repository, namespace, type)
+    }
+    var map = cachedMaps[namespace, type]
+    var unreconciledRepositories: Set<SingleNamespaceResourceRepository>? = null
+    if (map != null) {
+      unreconciledRepositories = unreconciledResources.get(namespace, type)
+      if (unreconciledRepositories == null) {
+        return map
+      }
+    }
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt:153]:
```suggestion
    if (map != null) {
      return map
    }
```

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt [hunk 12]
```diff
@@ -0,0 +1,535 @@
...
+
+    // Merge all items of the given type.
+    if (map == null) {
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt:156]:
```suggestion
```

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt [hunk 13]
```diff
@@ -0,0 +1,535 @@
...
+          unreconciledResources.keySet().toSet()
+        )
+        if (map.isEmpty) {
+          cachedMaps.remove(namespace, type)
+        }
+      }
+      unreconciledResources.remove(namespace, type)
+    }
+    return map
+  }
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt:215]:
```suggestion
```

unreconciled resources == psi synced resources, which is not needed here

**Review by jrodbx — CHANGES_REQUESTED** [review]

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/LocalResourceRepository.kt (hunk 3 above)
Comment (kevinzheng-ap) [reply paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/LocalResourceRepository.kt:44]: `getMapPackageAccessible` is used to in [MultiResourceRepository.getResources](https://github.com/cashapp/paparazzi/blob/9573caf8173df9170b90322c7adb3bf7b4fdfe44/paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/MultiResourceRepository.kt#L479). We can discuss tomorrow on which alternative function we could use

Comment (jrodbx) [reply paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/LocalResourceRepository.kt]: Ah, makes sense.  Thanks!

**Review by jrodbx — APPROVED** [review]
